File: tests/python/mistralai/test_openllmetry.py
"""Conformance test: OpenLLMetry (Traceloop) Mistral AI instrumentation.

Exercises: chat
against a mock OpenAI-compatible server, with the Traceloop Mistral AI instrumentation.
"""

import os

from otel_setup import setup_otel, flush_and_shutdown

MOCK_BASE_URL = os.environ["MOCK_LLM_URL"]


def instrument():
    from opentelemetry.instrumentation.mistralai import MistralAiInstrumentor
    MistralAiInstrumentor().instrument()


def run_chat(client):
    print("  [chat] basic chat completion")
    resp = client.chat.complete(
        model="mistral-large-latest",
        messages=[{"role": "user", "content": "Say hello."}],
    )
    print(f"    -> {resp.choices[0].message.content[:60]}")


def run_chat_streaming(client):
    print("  [chat_streaming] streaming chat completion")
    text = ""
    stream = client.chat.stream(
        model="mistral-large-latest",
        messages=[{"role": "user", "content": "Tell me a joke."}],
    )
    for event in stream:
        if event.data.choices and event.data.choices[0].delta.content:
            text += event.data.choices[0].delta.content
    print(f"    -> {text[:60]}")


def run_embeddings(client):
    print("  [embeddings] embedding generation")
    resp = client.embeddings.create(
        model="mistral-embed",
        inputs=["Hello, world!"],
    )
    print(f"    -> embedding dim: {len(resp.data[0].embedding)}")


def main():
    print("=== OpenLLMetry: Mistral AI Conformance Test ===")

    tp, lp, mp = setup_otel()
    instrument()

    from mistralai import Mistral
    client = Mistral(
        api_key="mock-key",
        server_url=MOCK_BASE_URL,
    )

    run_chat(client)
    try:
        run_chat_streaming(client)
    except Exception as e:
        print(f"    WARNING: streaming failed: {e}")

    run_embeddings(client)

    flush_and_shutdown(tp, lp, mp)


if __name__ == "__main__":
    main()

File: tests/js/langchain/test_openllmetry.ts
/**
 * Conformance test: OpenLLMetry (Traceloop) JS LangChain instrumentation.
 *
 * Exercises: chat, chat_streaming, embeddings
 * against a mock OpenAI server.
 */

import { LangChainInstrumentation } from "@traceloop/instrumentation-langchain";

import { flushAndShutdownOtel, setupOtel } from "../otel";

const MOCK_BASE_URL = process.env.MOCK_LLM_URL! + "/v1";

async function main() {
  console.log("=== OpenLLMetry JS: LangChain Conformance Test ===");

  const otel = setupOtel();

  // Dynamic import AFTER OTel setup so manuallyInstrument can patch prototypes.
  // tsx (ESM) does not support require-in-the-middle, so enable() does not work.
  const callbackManagerModule = await import("@langchain/core/callbacks/manager");
  const { ChatOpenAI } = await import("@langchain/openai");
  const { StringOutputParser } = await import("@langchain/core/output_parsers");
  const { ChatPromptTemplate } = await import("@langchain/core/prompts");

  const instrumentation = new LangChainInstrumentation();
  instrumentation.manuallyInstrument({ callbackManagerModule });

  const llm = new ChatOpenAI({
    model: "gpt-4o-mini",
    apiKey: "mock-key",
    configuration: { baseURL: MOCK_BASE_URL },
  });

  // Use RunnableSequence (pipe) so LangChain instrumentation can capture spans
  const prompt = ChatPromptTemplate.fromMessages([["user", "{input}"]]);
  const chain = prompt.pipe(llm).pipe(new StringOutputParser());

  console.log("  [chat] basic chat completion via chain");
  const resp = await chain.invoke({ input: "Say hello." });
  console.log(`    -> ${resp.slice(0, 60)}`);

  console.log("  [chat_streaming] streaming chat completion via chain");
  let text = "";
  const stream = await chain.stream({ input: "Tell me a joke." });
  for await (const chunk of stream) {
    text += chunk;
  }
  console.log(`    -> ${text.slice(0, 60)}`);

  console.log("  [embeddings] embedding generation");
  const { OpenAIEmbeddings } = await import("@langchain/openai");
  const embedModel = new OpenAIEmbeddings({
    model: "text-embedding-3-small",
    apiKey: "mock-key",
    configuration: { baseURL: MOCK_BASE_URL },
  });
  const embResult = await embedModel.embedQuery("Hello, world!");
  console.log(`    -> embedding dim: ${embResult.length}`);

  await flushAndShutdownOtel(otel);
}

main().catch((e) => { console.error(e); process.exit(1); });
